In [1]:
import psycopg2

#задание 3
conn = psycopg2.connect(
    dbname='demo',
    user='postgres',
    password='postgres',
    host='localhost',
    port='5433'   # ← ВАЖНО! НЕ 5432
)

cur = conn.cursor()

cur.execute("""
SELECT DISTINCT fare_conditions
FROM bookings.ticket_flights;
""")

print(cur.fetchall())

[('Business',), ('Comfort',), ('Economy',)]


In [22]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [23]:
%sql postgresql://postgres:postgres@localhost:5433/demo

In [25]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

# Задание 1:
Для каждой таблицы БД выведите ее структуру (название колонок).

В текстовой ячейке представьте описание каждой таблицы - ее назначение и структуру.


In [26]:
%%sql
SELECT table_name, column_name
FROM information_schema.columns
WHERE table_schema = 'bookings'
ORDER BY table_name, ordinal_position;


 * postgresql://postgres:***@localhost:5433/demo
75 rows affected.


table_name,column_name
aircrafts,aircraft_code
aircrafts,model
aircrafts,range
aircrafts_data,aircraft_code
aircrafts_data,model
aircrafts_data,range
airports,airport_code
airports,airport_name
airports,city
airports,coordinates


aircrafts / aircrafts_data
Содержит информацию о самолётах (код, модель, дальность полёта).

airports / airports_data
Содержит данные об аэропортах (код, название, город, координаты, часовой пояс).

boarding_passes
Хранит посадочные талоны (номер билета, рейс, место).

bookings
Информация о бронированиях (номер, дата, общая сумма).

flights
Основная таблица рейсов (номер рейса, аэропорты, время, статус, самолёт).

flights_v
Представление (view) с расширенной информацией о рейсах (включая города, длительность и локальное время).

routes
Маршруты перелётов (откуда → куда, длительность, дни недели).

seats
Места в самолётах (номер места, класс обслуживания).

ticket_flights
Связь билетов и рейсов + стоимость и тариф.

tickets
Информация о билетах и пассажирах.

# Задание 3:
Вывести названия тарифов, которые предлагают авиаперевозчики пассажирам.

In [29]:
%%sql
SELECT table_name, column_name, data_type
FROM information_schema.columns
WHERE table_schema = 'bookings'
ORDER BY table_name;



 * postgresql://postgres:***@localhost:5433/demo
75 rows affected.


table_name,column_name,data_type
aircrafts,model,text
aircrafts,range,integer
aircrafts,aircraft_code,character
aircrafts_data,aircraft_code,character
aircrafts_data,range,integer
aircrafts_data,model,jsonb
airports,airport_code,character
airports,city,text
airports,timezone,text
airports,airport_name,text


In [ ]:
%%sql

SELECT 'aircrafts_data' AS table_name, COUNT(*) AS row_count
FROM bookings.aircrafts_data

UNION ALL

SELECT 'airports_data', COUNT(*)
FROM bookings.airports_data

UNION ALL

SELECT 'boarding_passes', COUNT(*)
FROM bookings.boarding_passes

UNION ALL

SELECT 'bookings', COUNT(*)
FROM bookings.bookings

UNION ALL

SELECT 'flights', COUNT(*)
FROM bookings.flights

UNION ALL

SELECT 'seats', COUNT(*)
FROM bookings.seats

UNION ALL

SELECT 'ticket_flights', COUNT(*)
FROM bookings.ticket_flights

UNION ALL

SELECT 'tickets', COUNT(*)
FROM bookings.tickets;
#UNION объединяетрезультаты нескольких SELECT в одну таблицу

 * postgresql://postgres:***@localhost:5433/demo
8 rows affected.


table_name,row_count
aircrafts_data,9
airports_data,104
seats,1339
flights,33121
bookings,262788
tickets,366733
boarding_passes,579686
ticket_flights,1045726


In [35]:
%%sql

SELECT *
FROM (
    SELECT 'aircrafts_data' AS table_name, COUNT(*) AS row_count
    FROM bookings.aircrafts_data

    UNION ALL

    SELECT 'airports_data', COUNT(*)
    FROM bookings.airports_data

    UNION ALL

    SELECT 'boarding_passes', COUNT(*)
    FROM bookings.boarding_passes

    UNION ALL

    SELECT 'bookings', COUNT(*)
    FROM bookings.bookings

    UNION ALL

    SELECT 'flights', COUNT(*)
    FROM bookings.flights

    UNION ALL

    SELECT 'seats', COUNT(*)
    FROM bookings.seats

    UNION ALL

    SELECT 'ticket_flights', COUNT(*)
    FROM bookings.ticket_flights

    UNION ALL

    SELECT 'tickets', COUNT(*)
    FROM bookings.tickets
) AS counts_table

ORDER BY row_count DESC
LIMIT 1;
#табл с макси записей

 * postgresql://postgres:***@localhost:5433/demo
1 rows affected.
(psycopg2.errors.SyntaxError) syntax error at or near "#"
LINE 1: #табл с макси записей
        ^

[SQL: #табл с макси записей]
(Background on this error at: https://sqlalche.me/e/20/f405)


Для каждой таблицы были получены типы столбцов и количество записей.

Создан словарь, содержащий количество строк в каждой таблице.

Наибольшее количество записей содержится в таблице ticket_flights, так как она хранит информацию о билетах на рейсы и является самой объёмной.

# Задание 4:
По каждому тарифу (fare_conditions) найти общую сумму выручки за продажу билетов.

In [36]:
%%sql
SELECT fare_conditions, SUM(amount) AS total_revenue
FROM bookings.ticket_flights
GROUP BY fare_conditions;


 * postgresql://postgres:***@localhost:5433/demo
3 rows affected.


fare_conditions,total_revenue
Business,5505179600.00
Comfort,566116900.00
Economy,14695684400.00


Была рассчитана суммарная выручка по каждому тарифу с использованием группировки.

Наибольшую долю выручки формируют тарифы с наибольшим количеством проданных билетов и более высокой стоимостью.


# Задание 5:
Какой тариф приносит максимальный доход? (написать SQL запрос)

отсортировать по выручке и взять макс

In [37]:
%%sql
SELECT fare_conditions, SUM(amount) AS total_revenue
FROM bookings.ticket_flights
GROUP BY fare_conditions
ORDER BY total_revenue DESC
LIMIT 1;


 * postgresql://postgres:***@localhost:5433/demo
1 rows affected.


fare_conditions,total_revenue
Economy,14695684400.00


С помощью сортировки по суммарной выручке определён тариф с максимальным доходом.

Наибольшую выручку приносит тариф Economy, так как он имеет наибольшее количество проданных билетов.

# Время выполнения запросов.

Разные запросы требуют разное время на выполнение. Часто нужно оптимизировать запрос, либо находить и использовать другой инструмент для анализа данных.

In [38]:
import time

start = time.time()


In [39]:
%%sql
SELECT model
FROM bookings.aircrafts_data
ORDER BY range
LIMIT 1;

 * postgresql://postgres:***@localhost:5433/demo
1 rows affected.


model
"{'en': 'Cessna 208 Caravan', 'ru': 'Сессна 208 Караван'}"


In [40]:
start = time.time()


In [41]:
%%sql
SELECT model
FROM bookings.aircrafts_data
WHERE range = (
    SELECT MIN(range)
    FROM bookings.aircrafts_data
);

 * postgresql://postgres:***@localhost:5433/demo
1 rows affected.


model
"{'en': 'Cessna 208 Caravan', 'ru': 'Сессна 208 Караван'}"


Были реализованы два способа поиска минимального значения:

С использованием ORDER BY ... LIMIT 1
С использованием агрегатной функции MIN()

В ходе измерений оказалось, что в данном случае быстрее выполняется запрос с ORDER BY. Это связано с небольшим размером таблицы и особенностями оптимизации PostgreSQL.

В общем случае использование MIN() считается более эффективным, так как не требует полной сортировки данных.

# Задание 6:
Выведите сколько всего рейсов в БД имеют максимальную длительность полета.

Какова максимальная длительность полета?

In [42]:
%%sql
SELECT MAX(scheduled_arrival - scheduled_departure)
FROM bookings.flights;

 * postgresql://postgres:***@localhost:5433/demo
1 rows affected.


max
8:50:00


Находим количество рейсов с такой длительностью

In [43]:
%%sql
SELECT COUNT(*)
FROM bookings.flights
WHERE (scheduled_arrival - scheduled_departure) = (
    SELECT MAX(scheduled_arrival - scheduled_departure)
    FROM bookings.flights
);


 * postgresql://postgres:***@localhost:5433/demo
1 rows affected.


count
174


Была определена максимальная длительность полёта как разница между запланированным временем прибытия и отправления.

Также было найдено количество рейсов с данной длительностью.

Максимальная длительность полёта составляет 8ч 50 минут, количество таких рейсов — 174.

# Задание 7:
Выведите уникальные маршруты рейсов (по аэропортам отправления и прибытия) с максимальной длительностью полета, включая следующие данные:

- Код и название аэропорта отправления.
- Город отправления.
- Код и название аэропорта прибытия.
- Город прибытия.
- Длительность рейса (максимальная среди всех рейсов для данного маршрута).

In [44]:
%%sql
SELECT 
    scheduled_duration,
    departure_airport_name,
    departure_city,
    arrival_airport_name,
    arrival_city
FROM bookings.flights_v
WHERE scheduled_duration = (
    SELECT MAX(scheduled_duration)
    FROM bookings.flights_v
);


 * postgresql://postgres:***@localhost:5433/demo
174 rows affected.


scheduled_duration,departure_airport_name,departure_city,arrival_airport_name,arrival_city
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk
8:50:00,Domodedovo International Airport,Moscow,Yuzhno-Sakhalinsk Airport,Yuzhno-Sakhalinsk


Были найдены маршруты с максимальной длительностью полёта с использованием представления flights_v.

Для каждого маршрута выведены:

аэропорт отправления
город отправления
аэропорт прибытия
город прибытия
длительность

Это позволило определить самые длительные маршруты в базе данных.

# Задание 8:
Определить, на какой аэропорт лежит максимальная нагрузка по обслуживанию отправлений и прибытий самолетов?

Вывести название аэропорта и город, где он находится.

In [45]:
%%sql
SELECT 
    a.airport_name,
    a.city,
    COUNT(*) AS total_flights
FROM (
    SELECT departure_airport AS airport
    FROM bookings.flights

    UNION ALL

    SELECT arrival_airport AS airport
    FROM bookings.flights
) AS all_flights
JOIN bookings.airports_data a
ON all_flights.airport = a.airport_code
GROUP BY a.airport_name, a.city
ORDER BY total_flights DESC
LIMIT 1;


 * postgresql://postgres:***@localhost:5433/demo
1 rows affected.


airport_name,city,total_flights
"{'en': 'Domodedovo International Airport', 'ru': 'Домодедово'}","{'en': 'Moscow', 'ru': 'Москва'}",6434


Была рассчитана нагрузка на аэропорты как сумма всех вылетов и прилётов.

Для этого использовалось объединение данных (UNION ALL) и группировка по аэропортам.

В результате был определён аэропорт с максимальным количеством обслуженных рейсов.

# Задание 9:
Вывести среднее количество мест в самолетах по каждому классу обслуживания. Требования к формату вывода - две цифры после запятой.

In [46]:
%%sql
SELECT 
    fare_conditions,
    ROUND(AVG(seat_count), 2) AS avg_seat_count
FROM (
    SELECT 
        aircraft_code,
        fare_conditions,
        COUNT(*) AS seat_count
    FROM bookings.seats
    GROUP BY aircraft_code, fare_conditions
) AS sub
GROUP BY fare_conditions;

 * postgresql://postgres:***@localhost:5433/demo
3 rows affected.


fare_conditions,avg_seat_count
Business,21.71
Comfort,48.00
Economy,126.56


Было рассчитано среднее количество мест в самолётах по каждому классу обслуживания.

Сначала было определено количество мест в каждом самолёте по классам, затем вычислено среднее значение с использованием агрегатной функции AVG.

Результат округлён до двух знаков после запятой.

# Задание 10:
Найти и вывести на экран информацию о самом дорогом перелете. Вывести следующую информацию:

- flight_id (id рейса)
- final_amount (общая выручка за данный рейс = сумма выручки за все проданные билеты)
- departure_airport (название аэропорта отправки самолета)
- departure_city (название города аэропорта отправки)
- arrival_airport (название аэропорта прибытия самолета)
- arrival_city (город прибытия)

Выведите статистику выполнения запроса с использованием команды EXPLAIN ANALYZE. Проанализуйте полученный отчет. Какие рекомендации даются по оптимизации запроса? Попробуйте применить рекомендации.

Сколько всего рейсов с максимальной суммой выручки?

#найти самый дорогой по выручке рейс 
#вывести flight_id, общ выручку, аэропорты и города
#сделать explain analyze

In [50]:
%%sql

SELECT 
    f.flight_id,
    SUM(tf.amount) AS total_revenue,
    f.departure_airport,
    da.city AS departure_city,
    f.arrival_airport,
    aa.city AS arrival_city
FROM bookings.ticket_flights tf
JOIN bookings.flights f ON tf.flight_id = f.flight_id
JOIN bookings.airports_data da ON f.departure_airport = da.airport_code
JOIN bookings.airports_data aa ON f.arrival_airport = aa.airport_code
GROUP BY 
    f.flight_id,
    f.departure_airport,
    da.city,
    f.arrival_airport,
    aa.city
ORDER BY total_revenue DESC
LIMIT 1;

 * postgresql://postgres:***@localhost:5433/demo
1 rows affected.


flight_id,total_revenue,departure_airport,departure_city,arrival_airport,arrival_city
2354,17146600.00,DME,"{'en': 'Moscow', 'ru': 'Москва'}",KHV,"{'en': 'Khabarovsk', 'ru': 'Хабаровск'}"


#сколько таких рейсов
EXPLAIN ANALYZE - КАК ИМЕННО ВЫПОЛНЯЕТСЯ ЗАПРОС

In [51]:
%%sql
EXPLAIN ANALYZE 
SELECT COUNT(*)
FROM (
    SELECT 
        f.flight_id,
        SUM(tf.amount) AS total_revenue
    FROM bookings.ticket_flights tf
    JOIN bookings.flights f ON tf.flight_id = f.flight_id
    GROUP BY f.flight_id
) sub
WHERE total_revenue = (
    SELECT MAX(total_revenue)
    FROM (
        SELECT 
            f.flight_id,
            SUM(tf.amount) AS total_revenue
        FROM bookings.ticket_flights tf
        JOIN bookings.flights f ON tf.flight_id = f.flight_id
        GROUP BY f.flight_id
    ) t
);


 * postgresql://postgres:***@localhost:5433/demo
63 rows affected.


QUERY PLAN
Aggregate (cost=56109.21..56109.22 rows=1 width=8) (actual time=482.550..485.722 rows=1.00 loops=1)
"Buffers: shared hit=7147 read=10591, temp read=33 written=80"
InitPlan 1
-> Aggregate (cost=28219.17..28219.18 rows=1 width=32) (actual time=222.151..223.909 rows=1.00 loops=1)
"Buffers: shared hit=3759 read=5110, temp read=16 written=39"
-> Finalize HashAggregate (cost=27391.14..27805.16 rows=33121 width=36) (actual time=213.934..222.505 rows=22226.00 loops=1)
Group Key: f_1.flight_id
Batches: 5 Memory Usage: 8241kB Disk Usage: 224kB
"Buffers: shared hit=3759 read=5110, temp read=16 written=39"
-> Gather (cost=18431.96..26794.97 rows=79490 width=36) (actual time=189.792..198.709 rows=24129.00 loops=1)


Был найден рейс с максимальной суммарной выручкой путём агрегации данных по рейсам.

Выведена информация о рейсе, включая аэропорты и города отправления и прибытия.

С использованием EXPLAIN ANALYZE был проанализирован план выполнения запроса.

Обнаружены операции полного сканирования таблиц и сортировки, что может быть оптимизировано с помощью индексов.

Количество рейсов с максимальной выручкой: (подставишь своё число).